oauth2-cache-zio/src/main/scala/org/polyvariant/ocadotechnology/sttp/oauth2/cache/zio/CachingAccessTokenProvider.scala

```diff
@@ -0,0 +1,90 @@
+package org.polyvariant.sttp.oauth2.cache.zio
+
+import org.polyvariant.sttp.oauth2.AccessTokenProvider
+import org.polyvariant.sttp.oauth2.ClientCredentialsToken
+import org.polyvariant.sttp.oauth2.Secret
+import org.polyvariant.sttp.oauth2.cache.ExpiringCache
+import org.polyvariant.sttp.oauth2.cache.zio.CachingAccessTokenProvider.TokenWithExpirationTime
+import org.polyvariant.sttp.oauth2.common.Scope
+import zio.Clock
+import zio.Semaphore
+import zio.Task
+import zio.ZIO
+
+import java.time.Instant
+import scala.concurrent.duration.Duration
+
+final class CachingAccessTokenProvider(
+  delegate: AccessTokenProvider[Task],
+  semaphore: Semaphore,
+  tokenCache: ExpiringCache[Task, Option[Scope], TokenWithExpirationTime]
```

Comment:
question: if I understand correctly, this only works with an underlying `AccessTokenProvider` that has no environment. Wouldn't you want to support other kinds of [`RIO`](https://zio.dev/reference/core/zio/rio/)?

e.g.

```suggestion
final class CachingAccessTokenProvider[R](
  delegate: AccessTokenProvider[RIO[R, *]],
  semaphore: Semaphore,
  tokenCache: ExpiringCache[RIO[R, *], Option[Scope], TokenWithExpirationTime]
) extends AccessTokenProvider[RIO[R, *]]
```

Reply:
@Fristi wdyt?

Reply:
Good question. I think it would be nice to have it work with `RIO`, but would it be a common use case? These implementations don't need anything from the environment of RIO though. And in most cases I think (s|h)ttp clients are modelled as zio services where the dependency management is handled with `ZLayer` :)

Reply:
I'm not sure how common would that be, but better safe than binary breaking :wink: I turned `CachingAccessTokenProvider` to operate on `RIO`, exposed one constructor that takes an `R` and two others that still operate on `Task` for convenience. 

Please have another look and I guess we're good to merge and cut the release :rocket:

Reply:
TBH it's easy enough to implement a RIO wrapper on top of Task anyways 😅 you'd implement it as sth like `ZIO.env.flatMap(underlying.provide)`.

anyway, it's implemented with RIO now so I guess we can move on.